aicsimageio/readers/czi_reader.py

```diff
@@ -0,0 +1,795 @@
+import numpy as np
+import xarray as xr
+from dask import delayed
+from fsspec.spec import AbstractFileSystem
+
+from .. import constants, exceptions, metadata, types
+from ..dimensions import (
+    DEFAULT_CHUNK_DIMS,
+    DEFAULT_DIMENSION_ORDER_LIST_WITH_MOSAIC_TILES_AND_SAMPLES,
+    DEFAULT_DIMENSION_ORDER_LIST_WITH_SAMPLES,
+    REQUIRED_CHUNK_DIMS,
+    DimensionNames,
+)
+from ..utils import io_utils
+from .reader import Reader
+
+try:
+    from _aicspylibczi import BBox, TileInfo
+    from aicspylibczi import CziFile
+
+except ImportError:
+    raise ImportError(
+        "aicspylibczi is required for this reader. "
+        "Install with `pip install aicsimageio[czi]`"
+    )
+
+CZI_SAMPLES_DIM_CHAR = "A"
+CZI_BLOCK_DIM_CHAR = "B"
+CZI_SCENE_DIM_CHAR = "S"
+
+
+###############################################################################
+
+
+def _replace_sample_dim(dims: List[str]) -> List[str]:
+    return [
+        dim if dim != DimensionNames.Samples else CZI_SAMPLES_DIM_CHAR for dim in dims
+    ]
+
+
+###############################################################################
+
+
+DEFAULT_CZI_CHUNK_DIMS = _replace_sample_dim(DEFAULT_CHUNK_DIMS)
+REQUIRED_CZI_CHUNK_DIMS = _replace_sample_dim(REQUIRED_CHUNK_DIMS)
+DEFAULT_CZI_DIMENSION_ORDER_LIST = _replace_sample_dim(
+    DEFAULT_DIMENSION_ORDER_LIST_WITH_SAMPLES
+)
+DEFAULT_CZI_DIMENSION_ORDER_LIST_WITH_MOSAIC_TILES = _replace_sample_dim(
+    DEFAULT_DIMENSION_ORDER_LIST_WITH_MOSAIC_TILES_AND_SAMPLES
+)
+
+
+PIXEL_DICT = {
+    "gray8": np.uint8,
+    "gray16": np.uint16,
+    "gray32": np.uint32,
+    "bgr24": np.uint8,
+    "bgr48": np.uint16,
+}
+
+
+###############################################################################
+
+
+class CziReader(Reader):
+    """
+    Wraps the aicspylibczi API to provide the same aicsimageio Reader API but for
+    volumetric Zeiss CZI images.
+
+    Parameters
+    ----------
+    image: types.PathLike
+        Path to image file to construct Reader for.
+    chunk_by_dims: Union[str, List[str]]
+        Which dimensions to create chunks for.
+        Default: DEFAULT_CHUNK_DIMS
+        Note: Dimensions.SpatialY, Dimensions.SpatialX, and DimensionNames.Samples,
+        will always be added to the list if not present during dask array
+        construction.
+
+    Notes
+    -----
+    To use this reader, install with: `pip install aicsimageio[czi]`.
+    """
+
+    @staticmethod
+    def _is_supported_image(fs: AbstractFileSystem, path: str, **kwargs: Any) -> bool:
+        try:
+            with fs.open(path) as open_resource:
+                CziFile(open_resource)
+                return True
+
+        except RuntimeError:
+            return False
+
+    def __init__(
+        self,
+        image: types.PathLike,
+        chunk_by_dims: Union[str, List[str]] = DEFAULT_CHUNK_DIMS,
+    ):
+        # Expand details of provided image
+        self._fs, self._path = io_utils.pathlike_to_fs(image, enforce_exists=True)
```

Comment:
Right below this I would immediately check if the filepath provided was local or remote. Just like this: https://github.com/AllenCellModeling/aicsimageio/blob/main/aicsimageio/writers/ome_tiff_writer.py#L131